### ISSUE TITLE
Evaluation of `when` more lazily evaluates multiple conditions on one line.

### ISSUE THREAD

**Issue by marcandre** [jruby/jruby#6173]
DeepCover's test suite of semi-crazy corner cases found a regression in JRuby...

```
temp = [Symbol]; case :hello; when *temp, (p 'hola' || Integer); end
```

Should not print anything as the first condition is a match...

```
jruby 9.3.0.0-SNAPSHOT (2.6.5) 2020-04-13 d33953990e Java HotSpot(TM) 64-Bit Server VM 25.102-b14 on 1.8.0_102-b14 +jit [darwin-x86_64]
```

**Comment by enebo** [comment]
@marcandre do you know when this regressed?  Grammar changes happened on 2.6 support but I did not expect to see when affected.

**Comment by marcandre** [comment]
All I know is that it worked with jruby-head when [this build ran](https://travis-ci.org/github/deep-cover/deep-cover/builds/659020006) about a month ago, I can't be more precise.

**Comment by enebo** [comment]
@marcandre ok.  I will make sure this is not in 9.2.11.1 and it should just be a little more parser juggling...Thanks for the heads up.

**Comment by headius** [comment]
It does appear to be this way in 9.2.11 as well:

```
$ rvm jruby-9.2.11.0 do ruby -e "temp = [Symbol]; case :hello; when *temp, (p 'hola' || Integer); end"
"hola"
```

Here's the IR:

```
Instructions[-e#0#,IRScriptBody,AddMissingInitsPass]:
BB [1:LBL_3:-1] -EXIT->7 -FALL_THROUGH->2
	push_method_frame()
	push_method_binding()
BB [2:LBL_4:-1] -FALL_THROUGH->3 -REGULAR->4 -EXCEPTION->8
	%self = recv_self()
	line_num(;n: 0)
	%v_2 = search_const(%scope ;name: Symbol, no_priv: false)
	%v_3 = copy(Array:[%v_2])
	temp(0:0) = copy(%v_3)
	%v_6 = build_splat(%v_3)
	%v_8 = copy(strdup(frozen:"hola"))
	%v_7 = call_1o(%self, %v_8 ;n:p, t:FU, cl:false)
	%v_9 = build_compound_array(%v_6, %v_7 ;type: push)
	%v_5 = eqq(%v_9, :'hello' ;splat: true)
	b_true(LBL_2:15, %v_5)
BB [3:LBL_5:-1] -REGULAR->6 -EXCEPTION->8
	%v_4 = copy(nil)
	jump(LBL_0:21)
BB [4:LBL_2:15] -REGULAR->6 -EXCEPTION->8
	%v_4 = copy(nil)
	jump(LBL_0:21)
BB [6:LBL_0:21] -EXIT->7 -EXCEPTION->8
	pop_binding()
	pop_method_frame()
	return(%v_4)
BB [7:LBL_6:-1] -EXCEPTION->8
BB [8:_GLOBAL_ENSURE_BLOCK__0:-1] -EXIT->7
	%v_10 = recv_jruby_exc()
	pop_binding()
	pop_method_frame()
	throw(%v_10)
```

The AST indicates it's using ArgsPush here, constructing an array of all arguments to `when` before calling eqq:

```
[] ~/projects/jruby $ ast -e 'case a; when *b, (p "hola" || Integer); end'
AST:
RootNode line: 0
  CaseNode line: 0
    VCallNode:a line: 0
    ArrayNode line: 0
      WhenNode line: 0
        ArgsPushNode line: 0
          SplatNode line: 0
            VCallNode:b line: 0
          FCallNode:p line: 0
            ArrayNode line: 0
              OrNode line: 0
                StrNode line: 0
                ConstNode:Integer line: 0
, null
, , null
```

Perhaps this behavior changed in CRuby at some point?

**Comment by headius** [comment]
This is a Ruby 2.6 change (so 9.2.x is working properly):

```
[] ~/projects/jruby $ rvm ruby-1.9 do ruby -e "temp = [Symbol]; case :hello; when *temp, (p 'hola' || Integer); end"
"hola"

[] ~/projects/jruby $ rvm ruby-2.4.9 do ruby -e "temp = [Symbol]; case :hello; when *temp, (p 'hola' || Integer); end"
"hola"

[] ~/projects/jruby $ rvm ruby-2.5.7 do ruby -e "temp = [Symbol]; case :hello; when *temp, (p 'hola' || Integer); end"
"hola"

[] ~/projects/jruby $ rvm ruby-2.6.5 do ruby -e "temp = [Symbol]; case :hello; when *temp, (p 'hola' || Integer); end"

[] ~/projects/jruby $ 
```

**Comment by marcandre** [comment]
Right, sorry, I should have noticed that this was Ruby 2.6+ only

**Comment by enebo** [comment]
Going to fix this one today (most likely anyways).   MRI has logic when building that will build list, splat, argspush, and argscat in a method dedicated to compiling just when expressions.  For argspush (see AST above in @headius comment) we will buildWithOrder and that will end up calling buildArgsPush which will build both left and right hand side at the same time.  Those values end up getting stored as temps and used for the when eqq call.  In 2.6+, the lhs and rhs get built with some branches so that if the lhs matches then then rhs is never evaluated.

The fix to this problem is to make a buildWhenArgs method and do the logical equivalent to what MRI does.  A companion activity is to audit the other node types: splat, list, and argscat.

In looking ahead a bit list node is now completely destructured into more when cases in MRI.  This means:

  a. no List construction during evaluation
  b. More likely that in case of all literals we will optimize the case/when better.

**Comment by enebo** [comment]
This is fixed with #6738.
